---
title: "Proliferation & Serum-starved Medium Visualizations"
author: "Benji Bromberg"
output:
  toc: TRUE
---
# Background

Analysis of Mack1 cell RNA Seq from passage 27 (p27) and  passage 77 (p77) cells
in two cell states: proliferation and serum starvation

Investigation Goal:
What receptors are upregulated during growth and can we find ligands to aid 
  in serum-free media (SFM) development?
  
Outcomes using the results of this document:
Literature and vendor search for ligands to target the receptors. These ligands 
will be subsequently tested in serum-free media (SFM).

# Plans

Goals of this document:
--> PCA
    - filter expr_wide to only have p and s

--> Differential expression analysis between proliferation and serum-starvation
  Volcano plot (all genes, p27, p77)
    - sig genes diff color
    - sig receptors diff color
    
--> Pathway expression analysis between proliferation and serum-starvation
  Dot plot/bar plot (pathways.csv kegg.csv)
  - do for each ontology
  - filter adj.pval -> sort by log fold change
  - Michael examples
  
  Cluster heatmap (pathways.csv kegg.csv)
  - cluster by euclidiean distance

--> Identify upregulated receptors during proliferation
  Volcano plot (all genes)
    - sig genes diff color
    - sig receptors diff color
    - label receptors tested in culture
    - generate list of receptors

# Pipeline Scripts to Run
```{r}
# Libraries
source("General Pipeline Scripts/Libs for Gen Pipeline.R")

# Actual Processing
source("General Pipeline Scripts/Eggnog_Mapping.R")
source("General Pipeline Scripts/Data_Wrang_and_PreProcess.R")

# Functions
source("General Pipeline Scripts/DESeq2.R")
source("General Pipeline Scripts/GeneSymbol To GeneID.R")
```

# DE
```{r}
Output_protocol_p_s <- 
  (run_DESeq2(design_arg = ~ protocol, 
              contrast_arg = c("protocol", "p", "s"))
   )$df_res |> 
  add_GeneID_to_df()

Output_protocol_p77_p27 <- 
  (run_DESeq2(design_arg = ~ passage, 
              contrast_arg = c("passage", "p77", "p27"))
   )$df_res |> 
  add_GeneID_to_df()

Output_protocol_p27_p77 <- 
  (run_DESeq2(design_arg = ~ passage, 
              contrast_arg = c("passage", "p27", "p77"))
   )$df_res |> 
  add_GeneID_to_df()
```

# Receptor Information

Link to Receptor list:
https://www.ncbi.nlm.nih.gov/gene/?term=receptor%5BAll+Fields%5D+AND+%22thunnus+maccoyii%22%5Bporgn%5D+AND+alive%5Bprop%5D

  Search terms for NCBI genes:
    receptor[All Fields] AND "thunnus maccoyii"[porgn] AND alive[prop] 

## create_receptor_list function

Function to create a list of upregulated upregulated receptors from
the input csv file

```{r}
get_upreg_receptor_list <- function(csv_file_name) {
  # Load required packages
  library(dplyr)
  
  # Construct the full file path
  full_file_path <- paste0("output/dataframes/", csv_file_name)
  
  # Define a list of receptor genes to compare against
  receptor_list <- read.csv("thunnus_maccoyii_receptors.csv")
    # receptor_list is populated by a list of hits on the NCBI gene database
    # for the search "receptor" against the Thunnus Maccoyii genome.
  
  # Define differential gene expression data
  DGE <- read.csv(full_file_path)
  
  # Extract gene_id, LFC values, and adjusted p-values for each receptor gene
  receptors <- DGE %>% filter(gene_id %in% receptor_list$Symbol) %>% 
    select(gene_id, logFC, adj.P.Val)
  
  # Filter for significantly upregulated receptors 
    # (e.g., adjusted p-value < 0.05 and logFC > 1)
  upreg_receptors <- receptors %>% filter(adj.P.Val < 0.05, logFC > 1)
  
  # Return the top 15 upregulated receptors
  return(upreg_receptors)
}
```

## Create receptor lists
```{r}
upreg_receptors_p27 <- get_upreg_receptor_list("df_media_p27_genes.csv")
upreg_receptors_p77 <- get_upreg_receptor_list("df_media_p77_genes.csv")
upreg_receptors_combined <- get_upreg_receptor_list("df_media_combined_genes.csv")
```


# Volcano Plots
## create_volcano_plot function

```{r}
create_volcano_plot <- function(csv_file_name, plot_title, receptors_shown) {
  # Load required packages
  library(ggplot2)
  library(dplyr)
  library(ggrepel)
  
  # Define the significance threshold
  sig_threshold <- 0.05

  # Define the logFC threshold
  logFC_threshold <- 1
  
  # Construct the full file path
  full_file_path <- paste0("output/dataframes/", csv_file_name)
  
  # Define differential gene expression data
  DGE <- read.csv(full_file_path)
  
  
  # Create a new column to label significant genes
  DGE$significant <- ifelse(DGE$adj.P.Val <= sig_threshold & abs(DGE$logFC) >= logFC_threshold, "Significant", "Not Significant")

  if(receptors_shown == TRUE) {
    # Define the top receptors
    upreg_receptors <- get_upreg_receptor_list(csv_file_name)$gene_id
    
    # Create a new column to label upregulated receptors
    DGE$upreg_receptors <- ifelse(DGE$gene_id %in% upreg_receptors, "Upregulated Receptors", DGE$significant)
  
    # Set up the plot
    ggplot(DGE, aes(x = logFC, y = -log10(adj.P.Val), 
                    color = upreg_receptors)) +
      geom_point(size = 1) +
      
      # Add vertical line for logFC = 0
      geom_vline(xintercept = 0, linetype = "dashed", color = "gray40", 
                 size = 0.5) +
      
      # Add horizontal line for adj.P.Val = 0.05
      geom_hline(yintercept = -log10(0.05), linetype = "dashed", 
                 color = "gray40", size = 0.5) +
      
      # Add plot labels and title
      labs(x = "log2(Fold Change)", y = "-log10(Adjusted p-value)", 
           color = "", title = plot_title)
  } else {
    # Set up the plot
    ggplot(DGE, aes(x = logFC, y = -log10(adj.P.Val), 
                  color = significant)) + 
      geom_point(size = 1) + 
      
      # Add vertical line for logFC = 0
      geom_vline(xintercept = 0, linetype = "dashed", 
                 color = "gray40", size = 0.5) +
    
      # Add horizontal line for adj.P.Val = 0.05
      geom_hline(yintercept = -log10(0.05), linetype = "dashed", 
                 color = "gray40", size = 0.5) +
      
      # Add plot labels and title
      labs(x = "log2(Fold Change)", y = "-log10(Adjusted p-value)", 
           color = "", title = plot_title)
  }
}
```

## Create volcano plots
```{r}
volcano_p27 <- create_volcano_plot("df_media_p27_genes.csv", "Mack1 p27 DGE", FALSE)
volcano_p77 <- create_volcano_plot("df_media_p77_genes.csv", "Mack1 p77 DGE", FALSE)
volcano_combined <- create_volcano_plot("df_media_combined_genes.csv", "Mack1 DGE During Proliferation", TRUE)

# Display the plots
volcano_p27
volcano_p77
volcano_combined
```


# Cluster Heatmap

## create_heatmap function
```{r}
create_heatmap <- function(expr_data) {
  # Load required packages
  library(pheatmap)
  library(RColorBrewer)
  library(tidyverse)

  
  # Subset data to protocols p and s
  data_sub <- expr_data[, grepl("p|s", colnames(expr_data))]
  
  # Scale data to z-scores
  data_scale <- t(scale(t(data_sub)))
  
  #
  data_scale_sub <- as.data.frame(data_scale) |>
    select(-contains("m"), -contains("f")) |>
    as.matrix()
  
  # Create cluster heatmap
  pheatmap(data_scale_sub, 
           clustering_distance_rows = "euclidean",
           clustering_distance_cols = "euclidean",
           clustering_method = "complete")
}
```

```{r}
create_heatmap(expr_wide_cpm)
```


# Enrichment Pathways Visualization (version one pathways.csv)

## Bar plot: Least Significant Pathways (Filters: pval\<0.05; arrange(qvalue), top_n(30,qvalue))

```{r}
library(ggplot2)
library(dplyr)

# Load pathway enrichment results
enrichment_results <- read.csv("output/dataframes/df_media_p27_pathways.csv")

# Filter for significant pathways
enrichment_results <- enrichment_results %>%
  filter(p.adjust < 0.05)

# Sort by q-value
enrichment_results <- enrichment_results %>%
  arrange(qvalue) %>%
  top_n(30, qvalue)

# Create bar plot
ggplot(enrichment_results, aes(x = reorder(Term, -log10(qvalue)), y = -log10(qvalue), fill = Ontology)) +
  geom_col() +
  coord_flip() +
  xlab("Pathway") +
  ylab("-log10(q-value)") +
  ggtitle("Least Enriched Pathways in p27") +
  scale_fill_manual(values = c("#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3")) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) 

# Load pathway enrichment results
enrichment_results <- read.csv("output/dataframes/df_media_p77_pathways.csv")

# Filter for significant pathways
enrichment_results <- enrichment_results %>%
  filter(p.adjust < 0.05)

# Sort by q-value
enrichment_results <- enrichment_results %>%
  arrange(qvalue) %>%
  top_n(30, qvalue)

# Create bar plot
ggplot(enrichment_results, aes(x = reorder(Term, -log10(qvalue)), 
                               y = -log10(qvalue), fill = Ontology)) +
  geom_col() +
  coord_flip() +
  xlab("Pathway") +
  ylab("-log10(q-value)") +
  ggtitle("Least Enriched Pathways in p77") +
  scale_fill_manual(values = c("#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3")) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) 
```

## Bar plot: Most Significant Pathways (Filters: pval\<0.05; arrange(qvalue), top_n(-30,qvalue))

```{r}
library(ggplot2)
library(dplyr)

# Load pathway enrichment results
enrichment_results <- read.csv("output/dataframes/df_media_p27_pathways.csv")

# Filter for significant pathways
enrichment_results <- enrichment_results %>%
  filter(p.adjust < 0.05)

# Sort by q-value
enrichment_results <- enrichment_results %>%
  arrange(qvalue) %>%
  top_n(-30, qvalue)

# Create bar plot
ggplot(enrichment_results, aes(x = reorder(Term, -log10(qvalue)), y = -log10(qvalue), fill = Ontology)) +
  geom_col() +
  coord_flip() +
  xlab("Pathway") +
  ylab("-log10(q-value)") +
  ggtitle("Most Enriched Pathways in p27") +
  scale_fill_manual(values = c("#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3")) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) 

# Load pathway enrichment results
enrichment_results <- read.csv("output/dataframes/df_media_p77_pathways.csv")

# Filter for significant pathways
enrichment_results <- enrichment_results %>%
  filter(p.adjust < 0.05)

# Sort by q-value
enrichment_results <- enrichment_results %>%
  arrange(qvalue) %>%
  top_n(-30, qvalue)

# Create bar plot
ggplot(enrichment_results, aes(x = reorder(Term, -log10(qvalue)), y = -log10(qvalue), fill = Ontology)) +
  geom_col() +
  coord_flip() +
  xlab("Pathway") +
  ylab("-log10(q-value)") +
  ggtitle("Most Enriched Pathways in p77") +
  scale_fill_manual(values = c("#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3")) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) 
```

# Enrichment Pathways Visualization (version two)
## create_dotplot function (version 1)
```{r}
create_dotplot <- function(output) {
  # Load required packages
  library(enrichplot)
  
  
  # Filter the enriched pathways by adjusted p value and sort by log fold change
  kegg_df_filtered <- output$kegg_df %>%
    filter(p.adjust < 0.05) %>%
    arrange(desc(qvalue))
  
  bp_df_filtered <- output$pathways_df %>%
    filter(Ontology == "Biological Process", p.adjust < 0.05) %>%
    arrange(desc(qvalue))
  
  mf_df_filtered <- output$pathways_df %>%
    filter(Ontology == "Molecular Function", p.adjust < 0.05) %>%
    arrange(desc(qvalue))
  
  cc_df_filtered <- output$pathways_df %>%
    filter(Ontology == "Cellular Component", p.adjust < 0.05) %>%
    arrange(desc(qvalue))
  
  # Create the KEGG dotplot
  kegg_plot <- dotplot(medium_p27$genes_df, genes_df$qvalue, genes_df$adj.P.Val,
                      xlab = "Log Fold Change", pch = 19, cex = 0.8,
                      color = ifelse(kegg_df$qvalue > 0, "red", "blue"),
                      main = "KEGG Pathway Analysis")
  
  
  ggplot(kegg_df_filtered, aes(x = qvalue, y = category)) +
    geom_point(aes(size = -log10(p.adjust), color = qvalue > 0)) +
    scale_size_continuous(range = c(2, 8)) +
    scale_color_manual(values = c("tomato", "steelblue")) +
    labs(x = "Log Fold Change", y = "Category", 
         title = "KEGG Pathway Analysis") +
    theme_classic() +
    theme(legend.position = "right")
  
  # Create the BP dotplot
  bp_plot <- ggplot(bp_df_filtered, aes(x = qvalue, y = Term)) +
    geom_point(aes(size = -log10(p.adjust), color = qvalue > 0)) +
    scale_size_continuous(range = c(2, 8)) +
    scale_color_manual(values = c("tomato", "steelblue")) +
    labs(x = "Log Fold Change", y = "Term", 
         title = "GO Biological Process Analysis") +
    theme_classic() +
    theme(legend.position = "right")
  
  # Create the MF dotplot
  mf_plot <- ggplot(mf_df_filtered, aes(x = qvalue, y = Term)) +
    geom_point(aes(size = -log10(p.adjust), color = qvalue > 0)) +
    scale_size_continuous(range = c(2, 8)) +
    scale_color_manual(values = c("tomato", "steelblue")) +
    labs(x = "Log Fold Change", y = "Term", 
         title = "GO Molecular Function Analysis") +
    theme_classic() +
    theme(legend.position = "right")
  
  # Create the CC dotplot
  cc_plot <- ggplot(cc_df_filtered, aes(x = qvalue, y = Term)) +
    geom_point(aes(size = -log10(p.adjust), color = qvalue > 0)) +
    scale_size_continuous(range = c(2, 8)) +
    scale_color_manual(values = c("tomato", "steelblue")) +
    labs(x = "Log Fold Change", y = "Term", 
         title = "GO Cellular Component Analysis") +
    theme_classic() +
    theme(legend.position = "right")
  
  # Return a list of the four dotplots
  return(list(kegg_plot = kegg_plot, bp_plot = bp_plot, mf_plot = mf_plot,
              cc_plot = cc_plot))
}
```

## create_dotplot function (version 2)

```{r}
create_dotplot <- function(pathways_df, kegg_df) {
  # load required packages
  library(enrichplot)
  library(DOSE)
  
  # create dotplot for KEGG pathway data
  kegg_top30 <- kegg_df %>% 
    filter(p.adjust < 0.05) %>% 
    arrange(desc(Count)) %>% 
    head(30)

  kegg_dotplot <- enrichplot::dotplot(kegg_top30$category, 
                          -log10(kegg_top30$p.adjust), 
                          title = "KEGG Pathways", 
                          xlab = "-log10(Adjusted P Value)", 
                          pch = 16, 
                          cex = 0.8)

  # create dotplots for GO pathway data
  bp_top30 <- pathways_df %>% 
    filter(Ontology == "Biological Process", p.adjust < 0.05) %>% 
    arrange(desc(Count)) %>% 
    head(30)

  bp_dotplot <- enrichplot::dotplot(bp_top30$Term, 
                        -log10(bp_top30$p.adjust), 
                        title = "GO Biological Process", 
                        xlab = "-log10(Adjusted P Value)", 
                        pch = 16, 
                        cex = 0.8)

  mf_top30 <- pathways_df %>% 
    filter(Ontology == "Molecular Function", p.adjust < 0.05) %>% 
    arrange(desc(Count)) %>% 
    head(30)

  mf_dotplot <- enrichplot::dotplot(mf_top30$Term, 
                        -log10(mf_top30$p.adjust), 
                        title = "GO Molecular Function", 
                        xlab = "-log10(Adjusted P Value)", 
                        pch = 16, 
                        cex = 0.8)

  cc_top30 <- pathways_df %>% 
    filter(Ontology == "Cellular Component", p.adjust < 0.05) %>% 
    arrange(desc(Count)) %>% 
    head(30)

  cc_dotplot <- enrichplot::dotplot(cc_top30$Term, 
                        -log10(cc_top30$p.adjust), 
                        title = "GO Cellular Component", 
                        xlab = "-log10(Adjusted P Value)", 
                        pch = 16, 
                        cex = 0.8)

  # combine all dotplots into one figure
  grid.arrange(kegg_dotplot, bp_dotplot, mf_dotplot, cc_dotplot, ncol = 2)
}
```

You can call this function after running the `do_medium_analysis()` function for each medium condition:

```{r}
create_dotplot(medium_p27$pathways_df, medium_p27$kegg_df)
create_dotplot(medium_p77$pathways_df, medium_p77$kegg_df)
create_dotplot(medium_combined$pathways_df, medium_combined$kegg_df)
```

This will create four dotplots, one for each medium condition, showing the 30 most highly expressed genes in each KEGG pathway and GO ontology.

## Create dotplots
```{r}
create_dotplot(medium_p27)
```

To use the function, simply pass the output of `do_medium_analysis`

# Upregulated Receptor Analysis

GOAL: Use regular expression to identify receptors upregulated in proliferation media

receptor_list: https://www.ncbi.nlm.nih.gov/gene/?term=receptor%5BAll+Fields%5D+AND+%22thunnus+maccoyii%22%5Bporgn%5D+AND+alive%5Bprop%5D

```{r}
library(ggplot2)
library(ggrepel)

# Define a list of receptor genes to compare against
receptor_list <- read.csv("thunnus_maccoyii_receptors.csv")
  # receptor_list is populated by a list of hits on the NCBI gene database
  # for the search "receptor" against the Thunnus Maccoyii genome.

# Define 
medium_p27_genes <- read.csv("output/dataframes/df_media_p27_genes.csv")

# Extract gene_id, LFC values, and adjusted p-values for each receptor gene
medium_p27_receptors <- medium_p27_genes %>% 
  filter(gene_id %in% receptor_list$Symbol) %>% 
  select(gene_id, logFC, adj.P.Val)

# Filter for significantly upregulated receptors (e.g., adjusted p-value < 0.05 and logFC > 1)
upregulated_receptors <- medium_p27_receptors %>% 
  filter(adj.P.Val < 0.05, logFC > 1)

# Filter for the top 15 upregulated receptors by their adjusted p-value
top15_upregulated_receptors <- upregulated_receptors %>%
  top_n(-15, adj.P.Val)

# Visualize the results using a volcano plot
ggplot(medium_p27_receptors, aes(x = logFC, y = -log10(adj.P.Val), label = gene_id)) +
  geom_point(aes(color = ifelse(adj.P.Val < 0.05 & logFC > 1, "Upregulated", "Not upregulated")), alpha = 0.5) +
  scale_color_manual(values = c("Upregulated" = "red", "Not upregulated" = "black")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "gray50") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "gray50") +
  labs(x = "Log2-fold change", y = "-Log10(Adjusted P-Value)", color = "Condition") +
  ggtitle("Volcano plot of upregulated receptors in proliferation media") +
  theme_bw() + 
  geom_text_repel(data = top15_upregulated_receptors, aes(label = gene_id), nudge_x = 0.5, nudge_y = 0.5)
```

```{r}
library(ggplot2)
library(ggrepel)

# Define a list of receptor genes to compare against
receptor_list <- read.csv("thunnus_maccoyii_receptors.csv")
  # receptor_list is populated by a list of hits on the NCBI gene database
  # for the search "receptor" against the Thunnus Maccoyii genome.

# Define 
medium_p77_genes <- read.csv("output/dataframes/df_media_p77_genes.csv")

# Extract gene_id, LFC values, and adjusted p-values for each receptor gene
medium_p77_receptors <- medium_p77_genes %>% 
  filter(gene_id %in% receptor_list$Symbol) %>% 
  select(gene_id, logFC, adj.P.Val)

# Filter for significantly upregulated receptors (e.g., adjusted p-value < 0.05 and logFC > 1)
upregulated_receptors_p77 <- medium_p77_receptors %>% 
  filter(adj.P.Val < 0.05, logFC > 1)

# Filter for the top 15 upregulated receptors by their adjusted p-value
top15_upregulated_receptors_p77 <- upregulated_receptors_p77 %>%
  top_n(-15, adj.P.Val)

# Visualize the results using a volcano plot
ggplot(medium_p77_receptors, aes(x = logFC, y = -log10(adj.P.Val), label = gene_id)) +
  geom_point(aes(color = ifelse(adj.P.Val < 0.05 & logFC > 1, "Upregulated", "Not upregulated")), alpha = 0.5) +
  scale_color_manual(values = c("Upregulated" = "red", "Not upregulated" = "black")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "gray50") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "gray50") +
  labs(x = "Log2-fold change", y = "-Log10(Adjusted P-Value)", color = "Condition") +
  ggtitle("Volcano plot of upregulated receptors in serum-starved media") +
  theme_bw() + 
  geom_text_repel(data = top15_upregulated_receptors_p77, aes(label = gene_id), nudge_x = 0.5, nudge_y = 0.5)
```

```{r}
library(ggplot2)

# Read in the data from the CSV file
all_genes <- read.csv("output/dataframes/df_media_combined_genes.csv")

# Filter for the top 15 upregulated receptors by their adjusted p-value
top10_upregulated_genes <- all_genes %>%
  top_n(-10, adj.P.Val)

# Visualize the results using a volcano plot
ggplot(all_genes, aes(x = logFC, y = -log10(adj.P.Val), label = gene_id)) +
  geom_point(aes(color = ifelse(adj.P.Val < 0.05 & (logFC > 1 | logFC < -1), "Significant", "Not significant")), alpha = 0.3) +
  scale_color_manual(values = c("Significant" = "red", "Not significant" = "black")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "gray50") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "gray50") +
  labs(x = "Log2-fold change", y = "-Log10(Adjusted P-Value)", color = "Condition") +
  ggtitle("Volcano plot of genes in proliferation media") +
  theme_bw() + 
  geom_text_repel(data = top10_upregulated_genes, aes(label = gene_id), nudge_x = 0.5, nudge_y = 0.5)
```