---
title: "Proliferation & Serum-starved Medium Visualizations"
author: "Benji Bromberg"
output:
  toc: TRUE
---
# Background

Analysis of Mack1 cell RNA Seq from passage 27 (p27) and  passage 77 (p77) cells
in two cell states: proliferation and serum starvation

Investigation Goal:
What receptors are upregulated during growth and can we find ligands to aid 
  in serum-free media (SFM) development?
  
Outcomes using the results of this document:
Literature and vendor search for ligands to target the receptors. These ligands 
will be subsequently tested in serum-free media (SFM).

# Plans

Goals of this document:
--> PCA
    - filter expr_wide to only have p and s

--> Differential expression analysis between proliferation and serum-starvation
  Volcano plot (all genes, p27, p77)
    - sig genes diff color
    - sig receptors diff color
    
--> Pathway expression analysis between proliferation and serum-starvation
  Dot plot/bar plot (pathways.csv kegg.csv)
  - do for each ontology
  - filter adj.pval -> sort by log fold change
  - Michael examples
  
  Cluster heatmap (pathways.csv kegg.csv)
  - cluster by euclidiean distance

--> Identify upregulated receptors during proliferation
  Volcano plot (all genes)
    - sig genes diff color
    - sig receptors diff color
    - label receptors tested in culture
    - generate list of receptors

# Pipeline Scripts to Run
```{r}
setwd("~/Documents/Work/Kaplan/Mack RNA Seq")

# Libraries
source("General Pipeline Scripts/Libs for Gen Pipeline.R")

# Actual Processing
source("General Pipeline Scripts/Eggnog_Mapping (by GeneID).R")
source("General Pipeline Scripts/Data_Wrang_and_PreProcess.R")

# Functions
source("General Pipeline Scripts/DESeq2.R")
source("General Pipeline Scripts/GeneSymbol To GeneID.R")
source("General Pipeline Scripts/Pathway Enrichment (by GeneSymbol).R")
```

# DESeq2 Function Calls
```{r}
# Combined genes
Output_pro_combo <- 
  (run_DESeq2(design_arg = ~ protocol, 
              contrast_arg = c("protocol", "p", "s"))
   )$df_res |> 
  add_GeneID_to_df()

Output_pro_plus_pass_combo <- 
  (run_DESeq2(design_arg = ~ protocol + passage, 
              contrast_arg = c("protocol", "p", "s"))
   )$df_res |> 
  add_GeneID_to_df()

# Variation within p77
Output_pass_pro_p77 <- 
  (run_DESeq2(design_arg = ~ passage_protocol, 
              contrast_arg = c("passage_protocol", "p77_p", "p77_s"))
   )$df_res |> 
  add_GeneID_to_df()

# Variation within p27
Output_pass_pro_p27 <- 
  (run_DESeq2(design_arg = ~ passage_protocol, 
              contrast_arg = c("passage_protocol", "p27_p", "p27_s"))
   )$df_res |> 
  add_GeneID_to_df()
```

# Pathway Enrichment Function Calls
```{r}
# Combined genes
Pathway_pro_combo <- 
  run_pathway(Output_pro_combo$df_res_GeneID_for_pathway, 0.01)

Pathway_pro_plus_pass_combo <-  
  run_pathway(Output_pro_plus_pass_combo$df_res_GeneID_for_pathway, 0.01)

# Variation within p77
Pathway_pass_pro_p77 <- 
  run_pathway(Output_pass_pro_p77$df_res_GeneID_for_pathway, 0.01)

# Variation within p27
Pathway_pass_pro_p27 <- 
  run_pathway(Output_pass_pro_p27$df_res_GeneID_for_pathway, 0.01)
```

# Receptor Information

Link to Receptor list:
https://www.ncbi.nlm.nih.gov/gene/?term=receptor%5BAll+Fields%5D+AND+%22thunnus+maccoyii%22%5Bporgn%5D+AND+alive%5Bprop%5D

  Search terms for NCBI genes:
    receptor[All Fields] AND "thunnus maccoyii"[porgn] AND alive[prop] 

## get_upreg_receptor_list function

Function to create a list of upregulated upregulated receptors from
the input csv file

```{r}
GetUpregReceptorList <- function(df_res_GeneID) {
  # Set working directory
  setwd("~/Documents/Work/Kaplan/Mack RNA Seq/data")
  
  # Load required packages
  library(dplyr)
  
  # Define a list of receptor genes to compare against
  receptor_list <- read.csv("thunnus_maccoyii_receptors.csv")
    # receptor_list is populated by a list of hits on the NCBI gene database
    # for the search "receptor" against the Thunnus Maccoyii genome.
  
  upreg_receptors <- df_res_GeneID %>% 
    # Extract gene_id, LFC values, and adjusted p-values for each receptor gene.
    filter(gene_id %in% receptor_list$GeneID) %>% 
    select(gene_id, log2FoldChange, padj) %>% 
    
    # Filter for significantly upregulated receptors (e.g., adjusted 
    # p-value < 0.01 and log2FoldChange > 1
    filter(padj < 0.005, log2FoldChange > 2)
    
  # Add in metadata to receptor list and sort by log2FoldChange in 
  # decreasing order
  upreg_receptors_meta <- transform(upreg_receptors, 
                                    gene_id = as.numeric(gene_id)) %>%
    
    rename(GeneID = gene_id) %>%
    left_join(receptor_list, by = 'GeneID') %>%
    arrange(padj)
  
  # Return upregulated receptors list with metadata
  return(upreg_receptors_meta)
}
```

## Create lists of upregulated receptors

Additionally a document will be created for analysis
```{r}
# Lists of upregulated receptors for each condition
  # Combined genes
  Upreg_recep_pro_combo <- 
    GetUpregReceptorList(Output_pro_combo$df_res_GeneID)
  
  Upreg_recep_pro_plus_pass_combo <-  
    GetUpregReceptorList(Output_pro_plus_pass_combo$df_res_GeneID)
  
  # Variation within p77
  Upreg_recep_pass_pro_p77 <- 
    GetUpregReceptorList(Output_pass_pro_p77$df_res_GeneID)
  
  # Variation within p27
  Upreg_recep_pass_pro_p27 <- 
    GetUpregReceptorList(Output_pass_pro_p27$df_res_GeneID)

  # Write as csv files for investigation
  setwd("~/Documents/Work/Kaplan/Mack RNA Seq")
  write_csv(Upreg_recep_pro_combo,
            'Investigation 3/Receptor Lists/Upreg_recep_pro_combo.csv')
  write_csv(Upreg_recep_pro_plus_pass_combo,
            'Investigation 3/Receptor Lists/Upreg_recep_pro_plus_pass_combo.csv')
  write_csv(Upreg_recep_pass_pro_p77,
            'Investigation 3/Receptor Lists/Upreg_recep_pass_pro_p77.csv')
  write_csv(Upreg_recep_pass_pro_p27,
            'Investigation 3/Receptor Lists/Upreg_recep_pass_pro_p27.csv')

# Document outline for investigation of top 30 DEGs (ranked by padj)
  # Load required packages
  library(officer)
  library(flextable)
  
  # Create a new document object
  doc <- read_docx()
  
  # Create a bold style
  bold <- fp_text(bold = TRUE)
  not_bold <- fp_text(bold = FALSE)

  # Write document
  for(x in 30:1) {
    line1 <- as.character(Upreg_recep_pro_plus_pass_combo$GeneID[x])
    line2 <- Upreg_recep_pro_plus_pass_combo$Symbol[x]
    line3 <- Upreg_recep_pro_plus_pass_combo$description[x]
    line4 <- Upreg_recep_pro_plus_pass_combo$other_designations[x]
    line5 <- as.character(Upreg_recep_pro_plus_pass_combo$log2FoldChange[x])
    line6 <- as.character(Upreg_recep_pro_plus_pass_combo$padj[x])
    
    if(x != 30) {
      body_add_break(doc, pos="after")
    }
    
    # Prints headers in reverse order (prints upside down for some reason)
    
     body_add_fpar(doc, fpar(
      ftext("Findings: ", prop = bold), 
      ftext("", prop = not_bold)
    ))
    body_add_fpar(doc, fpar(
      ftext("", prop = not_bold) 
    ))
    body_add_fpar(doc, fpar(
      ftext("", prop = not_bold) 
    ))
    body_add_fpar(doc, fpar(
      ftext("PubMed Search: ", prop = bold) 
    ))
    body_add_fpar(doc, fpar(
      ftext("Adjusted p-value: ", prop = bold), 
      ftext(line6, prop = not_bold)
    ))
    body_add_fpar(doc, fpar(
      ftext("log2 Fold Change: ", prop = bold), 
      ftext(line5, prop = not_bold)
    ))
    body_add_fpar(doc, fpar(
      ftext("Other designations: ", prop = bold), 
      ftext(line4, prop = not_bold)
    ))
    body_add_fpar(doc, fpar(
      ftext("Description: ", prop = bold), 
      ftext(line3, prop = not_bold)
    ))
    body_add_fpar(doc, fpar(
      ftext("Symbol: ", prop = bold), 
      ftext(line2, prop = not_bold)
    ))
    body_add_fpar(doc, fpar(
      ftext("GeneID: ", prop = bold), 
      ftext(line1, prop = not_bold)
    ))
  }
  
  # Generate the Word document using the print function
  print(doc, target="Upregulated Receptor Analysis.docx")
```

# Volcano Plots
## create_volcano_plot function

```{r}
create_volcano_plot <- function(df_res_GeneID, plot_title, receptors_shown) {
  # Load required packages
  library(ggplot2)
  library(dplyr)
  library(ggrepel)
  
  # Define the significance threshold
  sig_threshold <- 0.01

  # Define the logFC threshold
  logFC_threshold <- 1
  
  # Define differential gene expression data
  DGE <- df_res_GeneID
  
  # Create a new column to label significant genes
  DGE$significant <- ifelse(DGE$padj <= sig_threshold & 
                              abs(DGE$log2FoldChange) >= logFC_threshold, 
                            "Significant", "Not Significant")

  if(receptors_shown == TRUE) {
    # Define the top receptors
    upreg_receptors <- GetUpregReceptorList(df_res_GeneID)$GeneID
    
    # Create a new column to label upregulated receptors
    DGE$upreg_receptors <- ifelse(DGE$GeneID %in% upreg_receptors, 
                                  "Upregulated Receptors", DGE$significant)
  
    # Set up the plot
    ggplot(DGE, aes(x = log2FoldChange, y = -log10(padj), 
                    color = upreg_receptors)) +
      geom_point(size = 1) +
      
      # Add vertical line for log2FoldChange = 0
      geom_vline(xintercept = 0, linetype = "dashed", color = "gray40", 
                 size = 0.5) +
      
      # Add horizontal line for padj = 0.01
      geom_hline(yintercept = -log10(0.01), linetype = "dashed", 
                 color = "gray40", size = 0.5) +
      
      # Add plot labels and title
      labs(x = "log2(Fold Change)", y = "-log10(Adjusted p-value)", 
           color = "", title = plot_title)
  } else {
    # Set up the plot
    ggplot(DGE, aes(x = log2FoldChange, y = -log10(padj), 
                  color = significant)) + 
      geom_point(size = 1) + 
      
      # Add vertical line for log2FoldChange = 0
      geom_vline(xintercept = 0, linetype = "dashed", 
                 color = "gray40", size = 0.5) +
    
      # Add horizontal line for padj = 0.01
      geom_hline(yintercept = -log10(0.01), linetype = "dashed", 
                 color = "gray40", size = 0.5) +
      
      # Add plot labels and title
      labs(x = "log2(Fold Change)", y = "-log10(Adjusted p-value)", 
           color = "", title = plot_title)
  }
}
```

## Create volcano plots
```{r}
volcano_combined <- 
  create_volcano_plot(Output_pro_combo$df_res_GeneID, 
                      "Mack1 DGE During Proliferation (~ Protocol)", 
                      TRUE)

volcano_combined_with_passage <- 
  create_volcano_plot(Output_pro_plus_pass_combo$df_res_GeneID, 
                      "Mack1 DGE During Proliferation (~ Protocol + Passage)", 
                      TRUE)

volcano_p77 <- 
  create_volcano_plot(Output_pass_pro_p77$df_res_GeneID, 
                      "Mack1 p77 DGE", 
                      FALSE)

volcano_p27 <- 
  create_volcano_plot(Output_pass_pro_p27$df_res_GeneID, 
                      "Mack1 p27 DGE", 
                      FALSE)

# Display the plots
volcano_combined
volcano_combined_with_passage
volcano_p77
volcano_p27
```


# Cluster Heatmap

## create_heatmap function
```{r}
create_heatmap <- function(expr_data) {
  # Load required packages
  library(pheatmap)
  library(RColorBrewer)
  library(tidyverse)

  
  # Subset data to protocols p and s
  data_sub <- expr_data[, grepl("p|s", colnames(expr_data))]
  
  # Scale data to z-scores
  data_scale <- t(scale(t(data_sub)))
  
  #
  data_scale_sub <- as.data.frame(data_scale) |>
    select(-contains("m"), -contains("f")) |>
    as.matrix()
  
  # Create cluster heatmap
  pheatmap(data_scale_sub, 
           clustering_distance_rows = "euclidean",
           clustering_distance_cols = "euclidean",
           clustering_method = "complete")
}
```

```{r}
create_heatmap(expr_wide)
```


#Bar Plots
```{r}

```

